src/metabase/models/cloud_migration.clj

```diff
@@ -0,0 +1,213 @@
+   [clj-http.client :as http]
+   [clojure.java.io :as io]
+   [clojure.set :as set]
+   [metabase.cmd.dump-to-h2 :as dump-to-h2]
+   [metabase.models.interface :as mi]
+   [metabase.models.setting :refer [defsetting]]
+   [metabase.models.setting.cache :as setting.cache]
+   [metabase.util.i18n :refer [deferred-tru tru]]
+   [metabase.util.log :as log]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]
+   [toucan2.pipeline :as t2.pipeline])
+  (:import
+   [java.io File InputStream]))
+
+(set! *warn-on-reflection* true)
+
+(def CloudMigration "Cloud Migration" :model/CloudMigration)
+
+(doto :model/CloudMigration
+  (derive :metabase/model)
+  (derive :hook/timestamped?))
+
+(methodical/defmethod t2/table-name :model/CloudMigration [_model] :cloud_migration)
+
+(t2/deftransforms :model/CloudMigration
+  {:state mi/transform-keyword})
+
+(defsetting metabase-store-migration-url
+  (deferred-tru "Store URL for migrations. Internal test use only.")
+  :visibility :internal
+  :default    "https://store-api.metabase.com/api/v2/migration"
+  :doc        false
+  :export?    false)
+
+(defsetting read-only-mode
+  (deferred-tru
+    (str "Boolean indicating whether a Metabase's is in read-only mode with regards to its app db. "
+         "Will take up to 1m to propagate to other Metabase instances in a cluster."
+         "Audit tables are excluded from read-only-mode mode."))
+  :type       :boolean
+  :visibility :admin
+  :default    false
+  :doc        false
+  :export?    false)
+
+(def ^:private read-only-mode-exceptions
+  #{;; Migrations need to update their own state
+    (t2/table-name :model/CloudMigration)
+    (t2/table-name :model/Setting)
+
+    ;; Users need to login, make queries, and we need need to audit them.
+    (t2/table-name :model/User)
+    (t2/table-name :model/Session)
+    (t2/table-name :model/LoginHistory)
+    (t2/table-name :model/AuditLog)
+    (t2/table-name :model/QueryExecution)
+    (t2/table-name :model/ViewLog)})
+
+(def ^:private ^:dynamic
+  *ignore-read-only-mode*
+  "Used during dump-to-h2, since rotate-encryption-key! over the dump will hit read-only-mode."
+  nil)
+
+;; Block write calls to most tables in read-only mode.
+(methodical/defmethod t2.pipeline/build :before [#_query-type     :toucan.statement-type/DML
+                                                 #_model          :default
+                                                 #_resolved-query :default]
+  [_query-type model _parsed-args resolved-query]
+  (when (and (read-only-mode)
+             (not (read-only-mode-exceptions (t2/table-name model)))
+             (not *ignore-read-only-mode*))
+    (throw (ex-info (tru "Metabase is in read-only-mode mode!")
+                    {:status-code 403})))
+  resolved-query)
+
+
+;; Helpers
+
+(def terminal-states
+  "Cloud migration states that are terminal."
+  #{:done :error :cancelled})
+
+(defn- cluster?
+  []
+  (>= (t2/count 'QRTZ_SCHEDULER_STATE) 2))
```

Comment:
this seems to work on h2 but on postgres:

```clojure
cloud-migration=> (metabase.db/db-type)
:postgres
cloud-migration=> (cluster?)
Execution error (PSQLException) at org.postgresql.core.v3.QueryExecutorImpl/receiveErrorResponse (QueryExecutorImpl.java:2725).
ERROR: relation "QRTZ_SCHEDULER_STATE" does not exist
  Position: 33
```

Changing this to `(t2/count 'qrtz_scheduler_state)` works on both h2 and postgres. Haven't checked mysql.

Reply:
Added a test that this should be 1 on tests, that should make the CI raise an error for the engines where it doesn't work.

Reply:
Looks like it's different just for Postgres, so added a conditional based on db-type
https://github.com/metabase/metabase/blob/0691779ddaf10f6713cdcd528f119c0f6392b056/resources/migrations/000_legacy_migrations.yaml#L4736-L4737